### PR TITLE
UCP/RNDV: Add rkey_ptr protocol

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Review by yosefe — COMMENTED** [review]
(publishing old comments)

File: src/ucp/core/ucp_request.h [hunk 1]
```diff
@@ -239,19 +239,29 @@ struct ucp_request {
...
+                                /* Used by rndv/send/ppln and rndv/recv/ppln */
+                                struct {
+                                    uint8_t send_ack;
+                                } ppln;
+
+                                /* Used by rndv/rkey_ptr */
+                                struct {
+                                    /* Element in queue for segmented RKEY ptr */
+                                    ucs_queue_elem_t  queue_elem;
+                                    /* Pointer for access to remote memory */
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:261]: space line before

File: src/ucp/proto/proto_common.c [hunk 2]
```diff
@@ -667,14 +667,16 @@ ucp_proto_common_init_caps(const ucp_proto_common_init_params_t *params,
     /* Calculate sender overhead */
     if (params->flags & UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY) {
         send_overhead = ucp_proto_common_memreg_time(params, reg_md_map);
-    } else {
+    } else if (!(params->flags & UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR)) {
```
Comment (yosefe) [inline src/ucp/proto/proto_common.c:670]: invert if/else to prefer positive condition over negative condition:
```
else if (params->flags & UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR){ 
  send_overhead = ucs_linear_func_make(0, 0);
} else {
  ucs_assert(reg_md_map == 0);
   ...
}
```

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR |
+                               UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY |
+                               UCP_PROTO_COMMON_INIT_FLAG_REMOTE_ACCESS |
+                               UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG,
+        .lane_type           = UCP_LANE_TYPE_RKEY_PTR,
+        .tl_cap_flags        = 0,
+    };
+
+    if ((init_params->select_param->op_id != UCP_OP_ID_RNDV_RECV) ||
+        (init_params->select_param->dt_class != UCP_DATATYPE_CONTIG)) {
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:46]: IMO ucp_datatype_iter_unpack should also support non-contig receive buffer

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    return ucp_proto_single_init(&params);
+}
+
+static unsigned ucp_proto_rndv_progress_rkey_ptr(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    ucp_request_t *req  = ucs_queue_head_elem_non_empty(&worker->rkey_ptr_reqs,
+                                                        ucp_request_t,
+                                                        send.rndv.rkey_ptr.queue_elem);
+    size_t seg_size     = worker->context->config.ext.rkey_ptr_seg_size;
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:59]: max_seg_size = ...

seg_size = ucs_min(max_seg_size, length - offset)

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    void *src;
+    int last;
+
+    seg_size   = ucs_min(seg_size, length - offset);
+    new_offset = offset + seg_size;
+    last       = new_offset == length;
+    src        = UCS_PTR_BYTE_OFFSET(req->send.rndv.rkey_ptr.ptr, offset);
+    status     = ucp_datatype_iter_unpack(&req->send.state.dt_iter, worker,
+                                          seg_size, offset, src);
+    req->send.rndv.offset = req->send.state.dt_iter.offset = new_offset;
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:73]: req->send.state.dt_iter.offset should not be used; should use req->send.state.completed_size

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    ucs_trace_data("rkey_ptr req: %p length: %zd seg: %zd off: %zd last: %d",
+                   req, length, seg_size, offset, last);
+    if (ucs_unlikely(status != UCS_OK) || last) {
+        ucs_queue_pull_non_empty(&worker->rkey_ptr_reqs);
+        ucp_proto_rndv_rkey_destroy(req);
+        ucp_proto_request_set_stage(req, UCP_PROTO_RNDV_RKEY_PTR_STAGE_ATS);
+        ucp_request_send(req);
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:80]: can unite common code with rndv_get?

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    ucs_queue_push(&worker->rkey_ptr_reqs, &req->send.rndv.rkey_ptr.queue_elem);
+    uct_worker_progress_register_safe(worker->uct,
+            ucp_proto_rndv_progress_rkey_ptr, worker,
+            UCS_CALLBACKQ_FLAG_FAST, &worker->rkey_ptr_cb_id);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_proto_rndv_rkey_ptr_ats_progress(uct_pending_req_t *uct_req)
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:121]: unite code with rndv_get?

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    ucp_request_t *req = ucs_container_of(uct_req, ucp_request_t, send.uct);
+
+    ucs_assert(ucp_datatype_iter_is_end(&req->send.state.dt_iter));
+    return ucp_proto_rndv_ack_progress(req, req->send.proto_config->priv,
+                                       UCP_AM_ID_RNDV_ATS,
+                                       ucp_proto_rndv_recv_complete);
+}
+
+static ucp_proto_t ucp_rndv_rkey_ptr_proto = {
+    .name       = "rndv/get/rkey_ptr",
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:132]: "rndv/rkey_ptr"

File: test/gtest/common/test_obj_size.cc [hunk 9]
```diff
@@ -56,7 +56,7 @@ UCS_TEST_F(test_obj_size, size) {
     EXPECTED_SIZE(ucp_rkey_t, 32);
 #endif
     /* TODO reduce request size to 240 or less after removing old protocols state */
-    EXPECTED_SIZE(ucp_request_t, 280);
+    EXPECTED_SIZE(ucp_request_t, 288);
```
Comment (yosefe) [inline test/gtest/common/test_obj_size.cc:59]: why needed to increase? can avoid?

File: src/ucp/core/ucp_request.h [hunk 10]
```diff
@@ -239,19 +239,29 @@ struct ucp_request {
...
+                                    void              *ptr;
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:262]: maybe it can be union with "mdesc" ?

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        .lane_type           = UCP_LANE_TYPE_RKEY_PTR,
+        .tl_cap_flags        = 0,
+    };
+
+    if ((init_params->select_param->op_id != UCP_OP_ID_RNDV_RECV) ||
+        (init_params->select_param->dt_class != UCP_DATATYPE_CONTIG)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_single_init(&params);
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:50]: check explicitly that have valid remote key in https://github.com/openucx/ucx/blob/f98ff97c7ed88bf1b7f3f3cbdee9ca28968e57e0/src/ucp/proto/proto_common.c#L314
- cannot skip for rkey_ptr proto, even if NEED_MEMH is false there

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    ucp_request_t *req                   = ucs_container_of(uct_req,
+                                                            ucp_request_t,
+                                                            send.uct);
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ucp_worker_h worker                  = req->send.ep->worker;
+    unsigned rkey_index                  = spriv->super.rkey_index;
+    ucp_rkey_h rkey                      = req->send.rndv.rkey;
+    ucs_status_t status;
+
+    status = uct_rkey_ptr(rkey->tl_rkey[rkey_index].cmpt,
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:102]: assert rkey_index is not -1

File: src/ucp/proto/proto_common.c [hunk 13]
```diff
@@ -311,15 +311,17 @@ static ucp_lane_index_t ucp_proto_common_find_lanes_internal(
...
-                    ucs_trace("%s: no support of dst md map 0x%" PRIx64,
-                              lane_desc, rkey_config_key->md_map);
-                    continue;
-                }
-            } else if (!(md_attr->cap.access_mem_types &
-                         UCS_BIT(rkey_config_key->mem_type))) {
+            if (((md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY) ||
+                 (flags & UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR)) &&
+                (!(rkey_config_key->md_map &
+                   UCS_BIT(ep_config_key->lanes[lane].dst_md_index)))) {
```
Comment (yosefe) [inline src/ucp/proto/proto_common.c:317]: extra ( )

File: src/ucp/rndv/rndv_rkey_ptr.c (hunk 3 above)
Comment (yosefe) [reply src/ucp/rndv/rndv_rkey_ptr.c:46]: unresolved

**Review by yosefe — APPROVED** [review]
pls fix merge conflict

**Comment by yosefe** [comment]
failure could be relevant
```
[ RUN      ] shm_ib_ipc/test_ucp_atomic64.post/5 <shm,ib,cuda_ipc,rocm_ipc,cuda_copy,rocm_copy/guess/proto>
[     INFO ] host->host ADD AND OR XOR 
[1632579118.289001] [swx-rdmz-ucx-gpu-01:11156:0]          amo_sw.c:227  UCX  ERROR Unsupported: got software atomic request while device atomics are selected on worker 0x49f0ad0
[swx-rdmz-ucx-gpu-01:11156:0:11156] Caught signal 11 (Segmentation fault: invalid permissions for mapped object at address 0x7f45dba00000)

/scrap/azure/agent-08/AZP_WORKSPACE/1/s/contrib/../src/ucp/rma/amo_sw.c: [ ucp_amo_sw_do_op64() ]
      ...
      201     }
      202 
      203 DEFINE_AMO_SW_OP(32)
==>   204 DEFINE_AMO_SW_OP(64)
      205 DEFINE_AMO_SW_FOP(32)
```

File: src/ucp/proto/proto_common.c [hunk 14]
```diff
@@ -311,15 +311,17 @@ static ucp_lane_index_t ucp_proto_common_find_lanes_internal(
...
+            if (((md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY) ||
+                 (flags & UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR)) &&
+                !(rkey_config_key->md_map &
+                  UCS_BIT(ep_config_key->lanes[lane].dst_md_index))) {
+                ucs_trace("%s: no support of dst md map 0x%" PRIx64,
+                          lane_desc, rkey_config_key->md_map);
+                continue;
+            }
+
+            if (!(md_attr->cap.access_mem_types &
```
Comment (yosefe) [inline src/ucp/proto/proto_common.c:323]:
```
if (!NEED_RKEY && 
    !(md_attr->cap.access_mem_types & UCS_BIT(rkey_config_key->mem_type)) {
  /* If no rkey - check the md can access the remote memory type */
```

File: src/ucp/proto/proto_common.c [hunk 15]
```diff
@@ -320,7 +320,8 @@ static ucp_lane_index_t ucp_proto_common_find_lanes_internal(
                 continue;
             }
 
-            if (!(md_attr->cap.access_mem_types &
+            if (!(md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY) &&
+                !(md_attr->cap.access_mem_types &
```
Comment (yosefe) [inline src/ucp/proto/proto_common.c:324]: pls add code comment here to explain and squash (without rebase)

Comment (Artemy-Mellanox) [reply src/ucp/proto/proto_common.c:326]: you mean add fixes  before merge commit?

**Review by yosefe — APPROVED** [review]

**Pull request merged**